### ISSUE TITLE
BPF+LLVM limitation: can't return a C struct by value

### ISSUE THREAD

**Issue by mvines** [solana-labs/solana#1802]
This sample BPF program:
```c
#include <solana_sdk.h>

struct foo {};
struct foo foo() {
  struct foo foo;
  return foo;
}

extern bool entrypoint(const uint8_t *input) {
  foo();
  return true;
}
```
fails to compile with llvm-7.0:
```
/usr/local/opt/llvm/bin/llc -march=bpf -filetype=obj  -o out/noop.o out/noop.bc
error: <unknown>:0:0: in function foo void (%struct.foo*): functions with VarArgs or StructRet are not supported
```

Seems like the function foo is unable to return the struct foo by value.

**Comment by jackcmay** [comment]
This is a similar error message seen when attempting to compile rust to BPF.

**Comment by jackcmay** [comment]
```rust
   if (IsVarArg || MF.getFunction().hasStructRetAttr()) {
     fail(DL, DAG, "functions with VarArgs or StructRet are not supported");
   }
```
http://llvm.org/doxygen/BPFISelLowering_8cpp_source.html

**Comment by jackcmay** [comment]
Fixed in #2100
